x/gamm/twap/logic_test.go

```diff
@@ -0,0 +1,206 @@
+package twap_test
+
+import (
+	"testing"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
+
+	"github.com/osmosis-labs/osmosis/v10/osmoutils"
+	"github.com/osmosis-labs/osmosis/v10/x/gamm/twap"
+	"github.com/osmosis-labs/osmosis/v10/x/gamm/twap/types"
+)
+
+var zeroDec = sdk.ZeroDec()
+var oneDec = sdk.OneDec()
+var twoDec = oneDec.Add(oneDec)
+var OneSec = sdk.NewDec(1e9)
+
+func newRecord(t time.Time, sp0, accum0, accum1 sdk.Dec) types.TwapRecord {
+	return types.TwapRecord{
+		Asset0Denom:     defaultUniV2Coins[0].Denom,
+		Asset1Denom:     defaultUniV2Coins[1].Denom,
+		Time:            t,
+		P0LastSpotPrice: sp0,
+		P1LastSpotPrice: sdk.OneDec().Quo(sp0),
+		// make new copies
+		P0ArithmeticTwapAccumulator: accum0.Add(sdk.ZeroDec()),
+		P1ArithmeticTwapAccumulator: accum1.Add(sdk.ZeroDec()),
+	}
+}
+
+// make an expected record for math tests, we adjust other values in the test runner.
+func newExpRecord(accum0, accum1 sdk.Dec) types.TwapRecord {
+	return types.TwapRecord{
+		Asset0Denom: defaultUniV2Coins[0].Denom,
+		Asset1Denom: defaultUniV2Coins[1].Denom,
+		// make new copies
+		P0ArithmeticTwapAccumulator: accum0.Add(sdk.ZeroDec()),
+		P1ArithmeticTwapAccumulator: accum1.Add(sdk.ZeroDec()),
+	}
+}
+
+func TestInterpolateRecord(t *testing.T) {
+	tests := map[string]struct {
+		record          types.TwapRecord
+		interpolateTime time.Time
+		expRecord       types.TwapRecord
+	}{
+		"0accum": {
+			record:          newRecord(time.Unix(1, 0), sdk.NewDec(10), zeroDec, zeroDec),
+			interpolateTime: time.Unix(2, 0),
+			expRecord:       newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+		},
+		"small starting accumulators": {
+			record:          newRecord(time.Unix(1, 0), sdk.NewDec(10), oneDec, twoDec),
+			interpolateTime: time.Unix(2, 0),
+			expRecord:       newExpRecord(oneDec.Add(OneSec.MulInt64(10)), twoDec.Add(OneSec.QuoInt64(10))),
+		},
+		"larger time interval": {
+			record:          newRecord(time.Unix(11, 0), sdk.NewDec(10), oneDec, twoDec),
+			interpolateTime: time.Unix(55, 0),
+			expRecord:       newExpRecord(oneDec.Add(OneSec.MulInt64(44*10)), twoDec.Add(OneSec.MulInt64(44).QuoInt64(10))),
+		},
+		"same time": {
+			record:          newRecord(time.Unix(1, 0), sdk.NewDec(10), oneDec, twoDec),
+			interpolateTime: time.Unix(1, 0),
+			expRecord:       newExpRecord(oneDec, twoDec),
+		},
+		// TODO: Overflow tests
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			// correct expected record based off copy/paste values
+			test.expRecord.Time = test.interpolateTime
+			test.expRecord.P0LastSpotPrice = test.record.P0LastSpotPrice
+			test.expRecord.P1LastSpotPrice = test.record.P1LastSpotPrice
+
+			gotRecord := twap.InterpolateRecord(test.record, test.interpolateTime)
+			require.Equal(t, test.expRecord, gotRecord)
+		})
+	}
+}
+
+func (s *TestSuite) TestUpdateTwap() {
```

Comment:
Some tests in this file utilize `TestSuite` while others don't. Let's make the use of test suite consistent please

Reply:
Why? Its fine for things that don't depend on app integration to just use testing.T ?

Reply:
Another purpose of the test suite is to group related spec together. I would assume that all the specs in the same file are logically grouped. 

In addition, just helps with consistency. I really like how we changed many of our test files post-audit to either use or not use the test suite

Reply:
These are logically grouped though? Unit testing logic files.

I think its _weirder_ to break them away by what interacts with state here.

In the gamm logic, I contrarily think that makes sense. I feel like this depends on file structure. These files aren't that big, making more files imo hurts readability.